File: eco-api/src/main/java/com/willfp/eco/util/ArrowUtils.java
package com.willfp.eco.util;

import org.bukkit.entity.Arrow;
import org.bukkit.inventory.ItemStack;
import org.bukkit.metadata.MetadataValue;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

/**
 * Utilities / API methods for arrows.
 */
public final class ArrowUtils {
    /**
     * Get the bow from an arrow.
     *
     * @param arrow The arrow.
     * @return The bow, or null if no bow.
     */
    @Nullable
    public static ItemStack getBow(@NotNull final Arrow arrow) {
        List<MetadataValue> values = arrow.getMetadata("shot-from");

        if (values.isEmpty()) {
            return null;
        }

        if (!(values.getFirst().value() instanceof ItemStack)) {
            return null;
        }

        return (ItemStack) values.getFirst().value();
    }

    private ArrowUtils() {
        throw new UnsupportedOperationException("This is a utility class and cannot be instantiated");
    }
}

File: eco-api/src/main/kotlin/com/willfp/eco/util/ArrowUtils.kt
@file:JvmName("ArrowUtilsExtensions")

package com.willfp.eco.util

import org.bukkit.entity.Arrow
import org.bukkit.inventory.ItemStack

/**
 * The bow that this arrow was shot from, read from the arrow's `shot-from` metadata.
 *
 * Null if the arrow has no such metadata (for example if it was not shot from a bow,
 * or if it was spawned by something that does not track the shooting item).
 *
 * @see ArrowUtils.getBow
 */
val Arrow.bow: ItemStack?
    get() = ArrowUtils.getBow(this)
